das/state.go

```diff
@@ -1,53 +1,234 @@
 package das
 
 import (
-	"sync"
-	"time"
+	"context"
 )
 
-// state collects information about the DASer process. Currently, there are
-// only two sampling routines: the main sampling routine which performs sampling
-// over current network headers, and the `catchUp` routine which performs sampling
-// over past headers from the last sampled checkpoint.
-type state struct {
-	sampleLk sync.RWMutex
-	sample   RoutineState // tracks information related to the main sampling routine
+// represents current state of sampling
+type coordinatorState struct {
+	rangeSize uint64
 
-	catchUpLk sync.RWMutex
-	catchUp   JobInfo // tracks information related to the `catchUp` routine
+	priority   []job                      // list of headers heights that will be sampled with higher priority
+	inProgress map[int]func() workerState // keeps track of running workers
+	failed     map[uint64]int             // stores heights of failed headers with amount of attempt as value
+
+	nextJobID int
+	next      uint64 // all headers before next were sent to workers
+	maxKnown  uint64 // max known header height
+
+	catchUpDone   bool          // indicates if all headers are sampled
+	catchUpDoneCh chan struct{} // blocks until all headers are sampled
+}
+
+func initSamplingState(samplingRangeSize uint64, c checkpoint) coordinatorState {
+	st := coordinatorState{
+		rangeSize:     samplingRangeSize,
+		inProgress:    make(map[int]func() workerState),
+		failed:        c.Failed,
+		next:          c.SampledBefore,
+		maxKnown:      c.MaxKnown,
+		catchUpDoneCh: make(chan struct{}),
+	}
+
+	if st.failed == nil {
+		st.failed = make(map[uint64]int)
+	}
+
+	st.priority = make([]job, 0, len(c.Failed)+len(c.Workers))
+	// put failed into priority to retry them on restart
+	for h := range c.Failed {
+		st.priority = append(st.priority, st.newJob(h, h, true))
+	}
+
+	// put last inProgress into priority
+	for _, wk := range c.Workers {
+		st.priority = append(st.priority, st.newJob(wk.From, wk.To, true))
+	}
+
+	if c.SampledBefore == 0 {
```

Comment:
It is possible if DASer was unable to load checkpoint from datastore and getting info from `Head()`

Reply:
```go
	// load latest DASed checkpoint
	cp, err := d.sampler.store.load(ctx)
	if err != nil {
		log.Warnw("checkpoint not found, initializing with height 1")

		cp = checkpoint{
			SampledBefore: 1,
			MaxKnown:      1,
		}

		if h, err := d.getter.Head(ctx); err == nil {
			cp.MaxKnown = uint64(h.Height)
		}
	}
	log.Info("starting DASer from checkpoint:\n", cp.String())

	d.sampler.state = initCoordinatorState(samplingRange, cp)
```

If the DASer is unable to load the checkpoint, it sets the checkpoint `cp` to

```
checkpoint{
			SampledBefore: 1,
			MaxKnown:      1,
		}
```

tries to load the Head and set the MaxKnown to head.Height, then passes this checkpoint into `initCoordinatorState`.

So how can `SampledBefore` be 0 then unless the node can load a checkpoint from disk that has `SampledBefore stored as 0?

-----------------------

Another point, in this [commit](https://github.com/celestiaorg/celestia-node/pull/988/commits/b7979991f3628404f62d3bf740079c1d1e0aab3f#diff-359d277b92395b8825cbffc2b991baa1dc4e1a9092f94ab3519feb30bb8bcee8R101), you again ignore the error from getter.Head: 

```
		if h, err := d.getter.Head(ctx); err == nil {
			cp.MaxKnown = uint64(h.Height)
		}
```

Why do you ignore it? [Header store is initialised](https://github.com/celestiaorg/celestia-node/blob/main/node/components.go#L89) when the node is started up, where it checks to see if it can load a head, if not, it initialises store's Head to the header of the config's trusted Hash. So if you get an error here, it is an issue.

Reply:
Sync discussion: Vlad is treating store kind of like a black box and doesn't "need" Head to operate the daser, which is why the error is ignored. We should probably doc it in the code though.

Reply:
I still do not know how `SampledBefore` can be 0.

Reply:
It can't. Previously it could be 0 between load and init funcs. I've moved setting default value for `SampledBefore` from `initCorrdinatorState` to `daser.Start` yesterday so `SampledBefore == 0` is impossible now

Reply:
I'll add comment to describe `Head()` invocation as opportunistic mechanic